rlp/list.libevm_test.go

```diff
@@ -54,3 +58,195 @@ func TestDecodeList(t *testing.T) {
+					&got,
+					[]any{obj.A, obj.B, obj.C},
+					[]any{obj.D, obj.E, obj.F},
+				)
+				require.NoErrorf(t, err, "EncodeStructFields(..., [required], [optional])")
+
+				assert.Equal(t, wantRLP, got.Bytes(), "EncodeToBytes() vs EncodeStructFields()")
+			})
+
+			t.Run("DecodeStructFields", func(t *testing.T) {
+				s := NewStream(bytes.NewReader(wantRLP), 0)
+				var got foo
+				err := s.DecodeStructFields(
+					[]any{&got.A, &got.B, &got.C},
+					[]any{&got.D, &got.E, &got.F},
+				)
+				require.NoError(t, err, "Stream.DecodeStructFields(...)")
+
+				var want foo
+				err = DecodeBytes(wantRLP, &want)
+				require.NoError(t, err, "DecodeBytes(...)")
+
+				assert.Equal(t, want, got, "DecodeBytes(..., [original struct]) vs Stream.DecodeStructFields(...)")
+			})
+		})
+	}
+}
+
+//nolint:testableexamples // Demonstrating code equivalence, not outputs.
+func ExampleStream_DecodeStructFields() {
+	type inner struct {
+		X uint64
+	}
+
+	type outer struct {
+		A uint64
+		B *inner `rlp:"nil"`
+		C *inner `rlp:"optional"`
+	}
+
+	val := outer{
+		A: 42,
+		B: &inner{X: 42},
+		C: &inner{X: 99},
+	}
+
+	// Errors are dropped for brevity for the sake of the example only.
+
+	_ = Encode(io.Discard, val)
+	// is equivalent to
+	_ = EncodeStructFields(
+		io.Discard,
+		[]any{val.A, val.B},
+		[]any{val.C},
+	)
+
+	r := bytes.NewReader(nil /*arbitrary RLP buffer*/)
+	var decoded outer
+	_ = Decode(r, &decoded)
+	// is equivalent to
+	_ = NewStream(r, 0).DecodeStructFields(
+		[]any{
+			&val.A,
+			Nillable(&val.B),
+		},
+		[]any{&val.C},
+	)
+
+	// Note the parallels between the arguments passed to
+	// {En,De}codeStructFields() and that, when decoding an optional or
+	// `rlp:"nil`-tagged field, a pointer to the _field_ is required even though
+	// in this example it will be a `**inner`.
+}
+
+func TestNillable(t *testing.T) {
+	type inner struct {
+		X uint64
+	}
+
+	type outer struct {
+		A *uint64   `rlp:"nil"`
+		B *inner    `rlp:"nil"`
+		C *[]uint64 `rlp:"nil"`
+	}
+
+	// Unlike the `rlp:"optional"` tag, there is no interplay between nil-tagged
+	// fields so we don't need the Cartesian product of all possible
+	// combinations.
+	var tests []outer
+	for _, a := range []*uint64{
+		nil,
+		common.PointerTo[uint64](0),
+	} {
+		tests = append(tests, outer{a, nil, nil})
+	}
+	for _, b := range []*inner{
+		nil,
+		{0},
+	} {
+		tests = append(tests, outer{nil, b, nil})
+	}
+	for _, c := range []*[]uint64{
+		nil,
+		{},
+		{0},
+	} {
+		tests = append(tests, outer{nil, nil, c})
+	}
+
+	// When a Nillable encounters an empty list it MUST set the field to nil,
+	// not just ignore it.
+	corruptInitialValue := func() outer {
+		return outer{common.PointerTo[uint64](42), &inner{42}, &[]uint64{42}}
+	}
+
+	for _, obj := range tests {
+		obj := obj
+		t.Run("", func(t *testing.T) {
+			rlp, err := EncodeToBytes(obj)
+			require.NoErrorf(t, err, "EncodeToBytes(%+v)", obj)
+			t.Logf("%s => %#x", pretty.Sprint(obj), rlp)
```

Comment:
nit remove perhaps

```suggestion
```

Reply:
It was very useful for debugging, effectively [identifying the input](https://google.github.io/styleguide/go/decisions#identify-the-input) for both the encoding and decoding steps.